compact_bytes/src/fixed.rs

```diff
@@ -0,0 +1,471 @@
+        }
+
+        length
+    }
+
+    /// Returns if the [`CompactBytesSlice`] is empty.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Returns the capacity of the [`CompactBytesSlice`].
+    #[inline(always)]
+    pub fn capacity(&self) -> usize {
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` share the same size and alignment.
+        //
+        // Note: This code is very carefully written so we can benefit from branchless
+        // instructions.
+        let (mut capacity, heap_capacity) = unsafe { (Self::MAX_INLINE, self.heap.len) };
+        if self.spilled() {
+            capacity = heap_capacity;
+        }
+        capacity
+    }
+
+    /// Consumes the [`CompactBytesSlice`], returning a `Box<[u8]>`.
+    #[inline]
+    pub fn into_boxed_slice(self) -> Box<[u8]> {
+        if self.spilled() {
+            // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+            // Above we checked if the current `CompactBytesSlice` is heap allocated.
+            let heap = unsafe { &self.heap };
+
+            let slice = unsafe { core::slice::from_raw_parts_mut(heap.ptr.as_ptr(), heap.len) };
+            let boxed = unsafe { Box::from_raw(slice) };
+            std::mem::forget(self);
+
+            boxed
+        } else {
+            // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+            // Above we checked if the current `CompactBytesSlice` is heap allocated.
+            let inline = unsafe { &self.inline };
+            inline.buffer.into()
+        }
+    }
+
+    /// Returns if the [`CompactBytesSlice`] has spilled to the heap.
+    #[inline(always)]
+    pub fn spilled(&self) -> bool {
+        // We store whether or not a `CompactBytesSlice` has spilled to the
+        // heap using the most significant bit of the last byte.
+        //
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+        unsafe { self.inline.data < INLINE_MASK }
+    }
+
+    #[inline(always)]
+    fn as_ptr(&self) -> *const u8 {
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+        //
+        // Note: This code is very carefully written so we can benefit from branchless
+        // instructions.
+        let mut pointer = self as *const Self as *const u8;
+        if self.spilled() {
+            pointer = unsafe { self.heap.ptr }.as_ptr()
+        }
+        pointer
+    }
+
+    #[inline(always)]
+    fn as_mut_ptr(&mut self) -> *mut u8 {
+        // SAFETY: `InlineBytes15` and `HeapBytesFixed` have the same size and alignment.
+        //
+        // Note: This code is very carefully written so we can benefit from branchless
+        // instructions.
+        let mut pointer = self as *mut Self as *mut u8;
+        if self.spilled() {
+            pointer = unsafe { self.heap.ptr }.as_ptr()
+        }
+        pointer
```

Comment:
Took me a moment to parse, maybe rewrite as an `if` in the return position with an `else` branch, here and above?

Reply:
Definitely a bit tricky to parse, but I wrote it specifically like this because it seemed to generate branchless accesses the most consistently